rfcs/0124-worker-manager.md

```diff
@@ -128,58 +135,176 @@ only information from the Queue stored in this service will be the task run
-from hardware".  Each provider a worker configuration can use would also have a
-section for provider-specific configuration.  This would store things like
-which worker image to use.  A more concrete example is that this section would
-store the AMI to use for an EC2 Worker.
-
-All levels of worker type configurations (e.g. global, per worker-type,
-per-provider) will support a `disabled: true` which will disable that section
-of the configuration.  The least specific disable will take priority.  This
-value will only block creation of new resources, not disable management of
-existing resources.
-
-Each edit of a worker-configuration will only be atomic to the
-worker-configuration level.
-
-### Multiple Worker Types for one Worker Configuration
-There are a lot of worker types which have identical configuration other than
-their worker type.  In order to support having more than one worker type for a
-given worker configuration, the worker configuration will have a list of worker
-types for which it can provide capacity for.  This will be a list in the
-configuration document.  Each entry in the list is either a string or an object
-which overrides global worker configuration options like max capacity for that
-specific worker type.  For configuration which is a range or value, the most
-restrictive will be used.
-
-Each worker type must be globally unique to a single worker configuration.
-
-### Documentation and Payload Schemata
+All JSON documents presented in this section will be considered illustrative of
+the idea and are not intended as a formal specification.
+
+Worker configurations will be a list of worker types and a list of rules
+specified in a JSON document.  Each rule will have conditions, values and a
+description.  The rules will be evaluated in the order of the rules list.
+
+Each rule set will be used to construct the internal configurations for each
+worker.  These configurations will be used by bidding strategies and providers
+as their native format.  The evaluation will begin with an empty object and
+each rule will write its values into that object.  Later rules can overwrite
+previous rules or can delete values from previous rules.
+
+This logic should be implemented as a standalone JavaScript library to allow
+reuse in tools which do automated worker configuration management as well as
+easier contributions to the rule set evaluation logic.
+
+Each edit of a worker configuration rule set will be atomic to precisely the
+worker configuration rule set.  Rules are intended to be building blocks rather
+than large blocks of configuration.  They must also be deterministic, such that
+they can be cached with invalidation only needed when a rule set is modified.
+
+Each worker type must be unique globally among all worker configurations.
+This will mean that any edits to a worker configuration which add new worker
+types will have to check all existing worker configurations to ensure there are
+no duplicated names.
```

Comment:
It looks like "workerType" is an input to the rules processing.  So somewhere outside of the rules there must be a map from workerType to worker configuration?

Reply:
when the manager would list worker types, or do anything which considers worker types, it would get the list of worker types from the worker configuration and execute the rules based on each worker type in the configuration.

Reply:
I suggested a worker type table earlier, since we have a many-to-one relationship between worker configurations and worker types.  Then the table structure itself can ensure that a worker type has exactly one configuration (because there's only one row for each worker type).  Is there any particular reason not to take this approach?